packages/frontend-plugin-api/src/components/ComponentRef.tsx

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  CoreBootErrorPageComponent,
+  CoreErrorBoundaryFallbackComponent,
+  CoreNotFoundErrorPageComponent,
+  CoreProgressComponent,
+} from '../types';
+
+/** @public */
+export type ComponentRef<T> = {
+  id: string;
+  T: T;
+};
+
+/** @public */
+export function createComponentRef<T>(options: {
+  id: string;
+}): ComponentRef<T> {
+  const { id } = options;
+  return {
+    id,
+    get T(): T {
+      throw new Error(`tried to read ComponentRef.T of ${id}`);
+    },
```

Comment:
Started avoiding this as it can be a pretty big annoyance in unit tests, could just skip and cast instead like here: https://github.com/backstage/backstage/blob/1cfe070efdd1bea7824442f6350e0b4cdd281d81/packages/frontend-plugin-api/src/wiring/createExtensionDataRef.ts#L48

Reply:
Done in a follow-up pull request: https://github.com/backstage/backstage/pull/21652.